src/compiler/builder.ts

```diff
@@ -1101,57 +1105,44 @@ namespace ts {
          * in that order would be used to write the files
          */
         function emit(targetSourceFile?: SourceFile, writeFile?: WriteFileCallback, cancellationToken?: CancellationToken, emitOnlyDtsFiles?: boolean, customTransformers?: CustomTransformers): EmitResult {
-            let restorePendingEmitOnHandlingNoEmitSuccess = false;

```

Comment:
This code was incorrect from the start.. The state of the world should be exactly same as using EmitAndSemanticDiagnsotics Builder as otherwise we have no guarantees of correctness.. The correct way to handle this is what we are doing now, that is removing files from pending emit based on emit done by the builder

Reply:
Thanks for the explanation!  Can you please put things like that in the description next time?